syntax = "proto3";

//生成多个java文件
option java_multiple_files = true;
//把生成的文件放到哪个包下
option java_package = "org.springblade.robot.grpc";
//输出的类名
option java_outer_classname = "RobotServiceProto";

import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";  // 用于任意JSON数据

// 服务定义
service RobotService {
    // 双向流式通信: 上传机器人状态、设备数据、环境数据、到达服务点位数据
    rpc clientUpload(stream RobotUploadRequest) returns (stream RobotUploadResponse);

    // 双向流式通信: 推送任务分配、服务指令
    rpc serverCommand(stream ClientStreamMessage) returns (stream ServerStreamMessage);
}

// 操作模式枚举
enum OperationMode {
    OPERATION_MODE_NONE = 0;
    OPERATION_MODE_OPEN_DOOR = 1;   // 开门
    OPERATION_MODE_CLOSE_DOOR = 2;  // 关门
    OPERATION_MODE_CAPTURE = 3; // 拍照
    OPERATION_MODE_SERVICE = 4; // 服务
}

// 站点状态枚举
enum StationTaskStatus {
    STATION_TASK_STATUS_UNSPECIFIED = 0;
    STATION_TASK_STATUS_PENDING = 1;      // 待执行
    STATION_TASK_STATUS_RUNNING = 2;      // 执行中
    STATION_TASK_STATUS_COMPLETED = 3;    // 已完成
    STATION_TASK_STATUS_FAILED = 4;       // 失败
    STATION_TASK_STATUS_CANCELLED = 5;    // 已取消
    STATION_TASK_STATUS_RETRYING = 6;     // 重试中
    STATION_TASK_STATUS_TO_RETRY = 7;     // 重试中
}

// 任务状态枚举
enum TaskStatus {
    TASK_STATUS_UNSPECIFIED = 0;
    TASK_STATUS_PENDING = 1;      // 待执行
    TASK_STATUS_RUNNING = 2;      // 执行中
    TASK_STATUS_COMPLETED = 3;    // 已完成
    TASK_STATUS_FAILED = 4;       // 失败
    TASK_STATUS_CANCELLED = 5;    // 已取消
    TASK_STATUS_RETRYING = 6;     // 重试中
}

// 消息类型
enum MsgType {
    ROBOT_STATUS = 0; // 机器人状态
    DEVICE_DATA = 1; // 巡检设备状态
    ENVIRONMENT_DATA = 2; // 巡检环境信息
    ARRIVE_SERVER_POINT = 3; // 到达服务器点通知
}

// 命令类型
enum CmdType {
    RESPONSE_CMD = 0; // 响应信息
    ROBOT_MODE_CMD = 1; // 机器人模式命令
    TASK_CMD = 2; // 任务下发命令
    JOY_CONTROL_CMD = 3; // 鼠标移动命令
    SET_MARKER_CMD = 4; // 设置标记点命令
    CHARGE_CMD = 5; // 充电命令
    POSITION_ADJUST_CMD = 6; // 重定位命令
    HEARTBEAT_CMD = 7; // 心跳确认
}

// 客户端信息类型
enum ClientMessageType {
    HEARTBEAT = 0; // 心跳
    COMMAND_RESPONSE = 1; // 普通命令响应
    SET_MARKER_RESPONSE = 2; // 设置标记点响应
}

// 机器人移动状态
enum MoveStatus {
    IDLE = 0; // 空闲状态，未接收移动指令
    RUNNING = 1; // 正在执行移动任务
    SUCCEEDED = 2; // 移动任务成功完成
    FAILED = 3; // 移动任务失败
    CANCELED = 4; // 移动任务被取消
    UNKNOWN = 5; // 未知 ADD
}

// 机器人模式
enum RobotMode {
    INSPECTION = 0; // 正常巡检模式
    SERVICE = 1; // 服务模式
    JOY_CONTROL = 2; // 手动控制模式
    ESTOP = 3; // 紧急停止模式
    CHARGE = 4; // 充电模式
    STAND_BY = 5; // 待命模式
    CONFIGURATION = 6; // 配置模式
}

//
// 上传相关
//

// 电池信息
message BatteryInfo {
    double power_percent = 1;            // 剩余电量百分比
    string charge_status = 2;    // 电池状态
}

// 位置信息
message PositionInfo {
    repeated double AGVPositionInfo = 1; // AGV位置 [x, y, theta]
    repeated double ARMPositionInfo = 2; // 机械臂关节位置 [J1-J6]
    repeated double EXTPositionInfo = 3; // 外部轴位置 [J1-J4]
}

// 任务信息
message TaskInfo {
    Task inspection_task_list = 1;
}

// 系统状态
message SystemStatus {
    MoveStatus move_status = 1;         // 机器人移动状态
    bool is_connected = 2;    // 通信连接状态
    bool soft_estop_status = 3;    // 软急停状态
    bool hard_estop_status = 4;    // 硬急停状态
    bool estop_status = 5;    // 急停状态（综合）
}

// 错误信息
message ErrorInfo {
    int32 code = 1;              // 错误编码
    string level = 2;            // 错误级别
    string message = 3;          // 错误信息
}

// 设备信息
message DeviceInfo {
    int64 device_id = 1;        // 设备ID
    string data_type = 2;        // 数据类型
    repeated string image_base64 = 3;     // 图片数据Base64
}

// 传感器数据
message EnvironmentInfo {
    double temperature = 1;      // 温度
    double humidity = 2;         // 湿度
    double oxygen = 3;           // O2
    double carbonDioxide = 4;   // CO2
    double pm25 = 5;             // PM2.5
    double pm10 = 6;             // PM10
    double etvoc = 7;            // eTVOC
    double noise = 8;            // 噪音
}

// 机器人状态上传
message RobotStatusUpload {
    BatteryInfo battery_info = 1;
    SystemStatus system_status = 2;
    ErrorInfo error_info = 3;
}


// 设备数据上传
message DeviceDataUpload {
    DeviceInfo device_info = 1;
}

// 环境数据上传
message EnvironmentDataUpload {
    EnvironmentInfo sensor_data = 1;
}

// 到达服务点位数据上传
message ArriveServicePointUpload {
    bool is_arrive = 1;
}

// 请求消息（用于双向流）
message RobotUploadRequest {
    int64 msg_id = 1;
    int64 msg_time = 2;
    MsgType msg_type = 3;
    int64 robot_id = 4;
    PositionInfo position_info = 5;
    TaskInfo task_info = 6;
    oneof data_json {
        RobotStatusUpload robot_status = 7;
        DeviceDataUpload device_data = 8;
        EnvironmentDataUpload environment_data = 9;
        ArriveServicePointUpload arrive_service_point = 10;
    }
}

// 响应信息
message ServerResponse {
    string code = 1;
    string info = 2;
}

// 请求消息响应
message RobotUploadResponse {
    int64 msg_id = 1;
    int64 msg_time = 2;
    MsgType msg_type = 3;
    int64 robot_id = 4;
    ServerResponse data_json = 5;
}

//
// 命令相关
//

// 操作配置
message OperationConfig {
    OperationMode operation_mode = 1;  // 操作模式
    string door_ip = 2;                // 门禁IP地址（可选）
    int64 device_id = 3;              // 设备ID（可选）
}

// 站点配置
message StationConfig {
    int64 station_id = 1;           // 站点ID（如"station1"）
    int32 sort = 2;                  // 站点执行顺序
    string name = 3;                 // 站点名称
    string agv_marker = 4;           // AGV导航点标识
    repeated double robot_pos = 5;   // 机械臂归位位置 [x, y, z, ...]
    repeated double ext_pos = 6;     // 外部轴归位位置 [x, y, z, ...]
    OperationConfig operation_config = 7; // 操作任务配置
}

// 站点信息
message Station {
    StationConfig station_config = 1;    // 站点配置
    StationTaskStatus status = 2;       // 任务状态，默认值在代码中处理
    int64 generate_time = 3;    // 任务生成时间
    int64 created_at = 4;    // 创建时间
    int64 started_at = 5;    // 开始时间（可选）
    int64 completed_at = 6;   // 完成时间（可选）
    int32 retry_count = 7;                       // 重试次数，默认值在代码中处理
    int32 max_retries = 8;                       // 最大重试次数，默认值在代码中处理
    string error_message = 9;                    // 错误信息（可选）
    map<string, google.protobuf.Value> meta_data = 10;         // 元数据，用于存储任意键值对
}

// 巡检任务
message Task {
    int64 task_id = 1; // 任务唯一标识符
    string task_name = 2; // 任务名称
    repeated Station station_list = 3; // 站点信息
    TaskStatus status = 4; // 任务状态，默认值在代码中处理
    RobotMode robot_mode = 5; // 执行任务时的机器人模式
    int64 generate_time = 6;    // 任务生成时间
    int64 created_at = 7;    // 创建时间
    int64 started_at = 8;    // 开始时间（可选）
    int64 completed_at = 9;   // 完成时间（可选）
    string error_message = 10; // 错误信息（可选）
    map<string, google.protobuf.Value> meta_data = 11; // 元数据，用于存储任意键值对
}


// 机器人模式命令
message RobotModeCommand {
    RobotMode robot_mode = 1;
}

// 机器人模式命令
message SetMarkerCommand {
    string marker_name = 1;
}


//鼠标控制移动
message JoyControlCmd {
    string angular_velocity = 1; // 角速度
    string linear_velocity = 2; // 线速度
}

//充电
message ChargeCmd {
    bool charge = 1; // 充电
}


// 服务端下发消息
message ServerStreamMessage {
    int64 command_id = 1;
    int64 command_time = 2;
    CmdType command_type = 3;
    int64 robot_id = 4;
    oneof data_json {
        RobotModeCommand robot_mode_command = 5;
        Task task_cmd = 6;
        JoyControlCmd joy_control_cmd = 7;
        ServerResponse response_info = 8;
        SetMarkerCommand set_marker_cmd = 9;
        ChargeCmd charge_cmd = 10;
    }
}

// 客户端消息
message ClientStreamMessage {
    int64 command_id = 1;
    int64 command_time = 2;
    ClientMessageType command_type = 3;
    int64 robot_id = 4;
    oneof data_json {
        ServerResponse response_info = 5;
        PositionInfo position_info = 6;
    }
}
